portfolio/src/main/java/com/google/sps/data/BookDaoDatastore.java

```diff
@@ -0,0 +1,206 @@
+        Set<String> genre = new HashSet<String>();
+        genre.addAll(genreList);
+        
+        Long tempPageCount = (Long) bookEntity.getProperty(PAGE_COUNT);
+        int pageCount = tempPageCount.intValue();
+ 
+        Book.Builder builder = Book.builder().title(title).genre(genre)
+            .categories(categories).authors(authors).language(language)
+            .description(description).infoLink(infoLink).pageCount(pageCount)
+            .publishedDate(publishedDate).publisher(publisher).thumbnail(thumbnail)
+            .maturityRating(maturityRating).isbn(isbn);
+ 
+        Book book = builder.build();
+        return book;
+
+    }
+
+    /**
+    * Parse an entity from a book
+    * @param bookEntity: the entity that you want to create a book from
+    */
+    private Entity parseEntity(Book book) {
+        //parse entity from book
+        long timeStamp = System.currentTimeMillis();
+
+        Entity bookEntity = new Entity(createKey(book.isbn()));
+
+        bookEntity.setProperty(TIME_STAMP,timeStamp);
+        bookEntity.setProperty(TITLE,book.title());
+        bookEntity.setProperty(GENRE,book.genre());
+        bookEntity.setProperty(CATEGORIES,book.categories());
+        bookEntity.setProperty(AUTHORS,book.authors());
+        bookEntity.setProperty(LANGUAGE,book.language());
+        bookEntity.setProperty(DESCRIPTION,book.description());
+        bookEntity.setProperty(INFO_LINK,book.infoLink());
+        bookEntity.setProperty(THUMBNAIL,book.thumbnail());
+        bookEntity.setProperty(PAGE_COUNT,book.pageCount());
+        bookEntity.setProperty(PUBLISHED_DATE,book.publishedDate());
+        bookEntity.setProperty(PUBLISHER,book.publisher());
+        bookEntity.setProperty(MATURITY_RATING,book.maturityRating());
+        bookEntity.setProperty(ISBN,book.isbn());
+
+        return bookEntity;
+    }
+
+    /**
+    * Creates a custom Key object of kind ENTITY_KIND ("Book")
+    * @param isbn: The isbn of the book you want to create the key for
+    * the same isbn will create the same key so it will work when
+    * passing isbn to get book
+    */
+    private Key createKey(String isbn) {
+        return KeyFactory.createKey(ENTITY_KIND, isbn);
+    }
+
+    /**
+    * Checks to see if the input is valid
+    * @param obj: The input you want to 
+    * validate (isbn or book)
+    */
+    private boolean validate(Object obj){
+        if(obj == null){
+            throw new NullPointerException("The input was null.");
+        }
+
+        String isbn = null;
+        if(obj.getClass().equals(java.lang.String.class)){
+            isbn = (String) obj;
+        }else if(obj.getClass().equals(com.google.sps.data.AutoValue_Book.class)){
+            Book book = (Book) obj;
+            isbn = book.isbn();
+        }
+
+        if(isbn != null && isbn.matches("[0-9]+") && isbn.length() == 13){
+            return true;
+        }
+        throw new RuntimeException(String.format("Bad argument %s", obj));
+    }
```

Comment:
You should create separate methods for validations: validateIsbn validateBook.

Optional: for some of these things, please try to use Guava preconditions: https://guava.dev/releases/19.0/api/docs/com/google/common/base/Preconditions.html